### PR TITLE
Deprecate `cat(dim, Xs...)` to `cat(Xs..., dims=dim)`

### LINKED ISSUE #27100: Change `cat` to use `dims` keyword arguments
It was pointed out that `cat` wasn't included in the recent refactor to make
dimensionality a keyword argument. `cat` is a bit different than the other such
functions in that it is often called with only a single dimension, but the argument does
generalize:
```
julia> cat(1, 2, 3, 4)
3-element Array{Int64,1}:
 2
 3
 4

julia> cat((1,2), 2, 3, 4)
3×3 Array{Int64,2}:
 2  0  0
 0  3  0
 0  0  4
```
Given those, examples, I think it would be nice to explicitly call out which argument is the dimension, e.g.:
```
cat(2, 3, 4; dims = 1)
```
There's a question what the default should be if no argument is passed. For now, we can't really do anything abut that because of the deprecation (so we should make calls that don't match the deprecation and don't have the keyword arg an error), but it may be nice to eventually do something automatic based on the argument types.

### PR DESCRIPTION
**Pull request by mbauman**
Fixes #27100.  I have also started on a bigger refactor here that will more easily allow for custom arrays to extend concatenations, but that ended up being not as simple as I had hoped, so this PR simply moves to the `dims` keyword argument and hides all internal dispatch so Base only exposes one `cat` method without restrictions on its arguments for now.

### CODE REVIEW

**Comment by KristofferC** [comment]
32bit CI loads for me and got:

```
 Got exception ErrorException("Process timed out possibly waiting for a response. Process output found:\n\"\"\"\n\r\nsignal (15): Terminated\r\nin expression starting at no file:0\r\n_ZN4llvm13ConstantRange21makeAllowedICmpRegionENS_7CmpInst9PredicateERKS0_ at /tmp/julia/bin/../lib/julia/libLLVM-6.0.so (unknown line)\r\n_ZN4llvm13ConstantRange24makeSatisfyingICmpRegionENS_7CmpInst9PredicateERKS0_ at /tmp/julia/bin/../lib/julia/libLLVM-6.0.so (unknown line)\r\n_ZN4llvm13ConstantRange19makeExactICmpRegionENS_7CmpInst9PredicateERKNS_5APIntE at /tmp/julia/bin/../lib/julia/libLLVM-6.0.so (unknown line)\r\n\n\"\"\"") outside of a @test
  Process timed out possibly waiting for a response. Process output found:
  """
  

  signal (15): Terminated

  in expression starting at no file:0

  _ZN4llvm13ConstantRange21makeAllowedICmpRegionENS_7CmpInst9PredicateERKS0_ at /tmp/julia/bin/../lib/julia/libLLVM-6.0.so (unknown line)

  _ZN4llvm13ConstantRange24makeSatisfyingICmpRegionENS_7CmpInst9PredicateERKS0_ at /tmp/julia/bin/../lib/julia/libLLVM-6.0.so (unknown line)

  _ZN4llvm13ConstantRange19makeExactICmpRegionENS_7CmpInst9PredicateERKNS_5APIntE at /tmp/julia/bin/../lib/julia/libLLVM-6.0.so (unknown line)

  
  """
  Stacktrace:
   [1] error(::String, ::String) at ./error.jl:42
   [2] (::getfield(Main.Test36Main_LibGit2_libgit2.LibGit2Tests, Symbol("##5#8")){Int32,Cmd,Array{Any,1},getfield(Main.Test36Main_LibGit2_libgit2.LibGit2Tests, Symbol("#format_output#7")){Bool},Base.GenericIOBuffer{Array{UInt8,1}}})(::RawFD, ::Base.TTY) at /tmp/julia/share/julia/stdlib/v0.7/LibGit2/test/libgit2.jl:93
   [3] with_fake_pty(::getfield(Main.Test36Main_LibGit2_libgit2.LibGit2Tests, Symbol("##5#8")){Int32,Cmd,Array{Any,1},getfield(Main.Test36Main_LibGit2_libgit2.LibGit2Tests, Symbol("#format_output#7")){Bool},Base.GenericIOBuffer{Array{UInt8,1}}}) at /tmp/julia/share/julia/test/TestHelpers.jl:37
...
```

**Comment by Keno** [comment]
I think we may have to increase those timeouts for libgit2

**Comment by JeffBezanson** [comment]
Note on the merge conflict: I moved this code to util.jl (with the rest of the text-color-related code, which I did to resolve a bootstrap order thing, but it makes sense anyway).

File: doc/src/manual/arrays.md [hunk 1]
```diff
@@ -86,8 +86,8 @@ Arrays can be constructed and also concatenated using the following functions:
 | Function               | Description                                          |
 |:---------------------- |:---------------------------------------------------- |
 | [`cat(k, A...)`](@ref) | concatenate input n-d arrays along the dimension `k` |
```
Comment (i-apellaniz) [inline doc/src/manual/arrays.md]: > `| [\`cat(k, A...)\`](@ref) ... `
it should be `cat(A..., dims = k)`, isn't it?

**Pull request merged**